core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

```diff
@@ -104,6 +104,12 @@ class SparkHadoopUtil extends Logging {
       }
       val bufferSize = conf.get("spark.buffer.size", "65536")
       hadoopConf.set("io.file.buffer.size", bufferSize)
+
+      if (conf.contains("spark.sql.default.derby.dir")) {
```

Comment:
@yhuai 
Spark by default has derby for metastore. Generally metastore_db and derby.log gets created by default in the current directory. This creates a problem for more restrictive environment, such as when running as a R package when the guideline is not to have anything written to user's space (unless under tempdir)

Just checking now it also seems to be the case when running the pyspark shell.

It looks like this is the new behavior since 2.0.0. Would it make sense if we always default derby/metastore to tempdir unless it is running in an application directory that would be cleaned out when the job is done (eg. YARN cluster)